File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int best = 0;

        int l = 0;
        int r = 0;
        
        unordered_set<char> seen;

        while (r < s.size()) {

            if (seen.insert(s[r]).second) {
                r++;
            } else {
                seen.erase(s[l]);
                l++;
            }
            best = max(best, (int)seen.size());
        }
        return best;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-7.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        l = 0
        r = 1

        minval = 0
        best = 0
        # 3, 5, 1, 10
        while r < len(prices):

            if prices[r] < prices[l]:
                l = r
            else:
                best = max(best, prices[r] - prices[l])
            
            r = r + 1
        
        return best